tools/python/boutiques/puller.py

```diff
@@ -0,0 +1,77 @@
+import requests
+import urllib
+try:
+    # Python 3
+    from urllib.request import urlopen
+    from urllib.request import urlretrieve
+except ImportError:
+    # Python 2
+    from urllib2 import urlopen
+    from urllib import urlretrieve
+
+
+class ZenodoError(Exception):
+    pass
+
+
+class Puller():
+
+    def __init__(self, identifier, verbose, download):
+        self.identifier = identifier
+        self.verbose = verbose
+        self.download = download
+
+    def pull(self):
+        r = requests.get('https://zenodo.org/api/records/?q=boutiques&'
+                         'keywords=boutiques&keywords=schema&'
+                         'keywords=version&file_type=json&type=software')
+
+        if(r.status_code != 200):
+            self.raise_zenodo_error("Error searching Zenodo", r)
+
+        try:
+            int(self.identifier)
+            return self.pull_by_id(r, self.identifier)
+        except ValueError:
+            return self.pull_by_filename(r, self.identifier)
+
+    def pull_by_id(self, r, zid):
+        for hit in r.json()["hits"]["hits"]:
+            file_path = hit["files"][0]["links"]["self"]
+            file_name = file_path.split("/")[-1]
+            if hit["id"] == int(zid):
+                if self.download:
+                    if(self.verbose):
+                        self.print_zenodo_info("Downloading descriptor %s"
+                                               % file_name, r)
+                    return urlretrieve(file_path, file_name)
```

Comment:
It would be nice to store the descriptors in some common place, for instance under `.boutiques` (but be careful, `.boutiques` is currently a file containing Zenodo tokens), so that they can be cached. If descriptors are immutable, caching won't create additional issues. Caching would also allow to work offline seamlessly or be robust to Zenodo downtimes. It should be easy enough to do, but feel free to add this as an issue for future work.

Reply:
I looked into this and it seems like it will cause problems with the publisher, which loads the entire `.boutiques` file to read the access tokens. Rather than refactoring the publisher to accommodate this, what if we cached the descriptors in the user's `.cache` folder?

Reply:
You are right, using `.cache` makes much more sense! Note that `os.getenv("HOME")` wouldn't work on Windows (which we don't support yet but might in the future). Here is how to get the HOME cross-platform: https://stackoverflow.com/questions/2668909/how-to-find-the-real-user-home-directory-using-python If you're doing it, could you also update the publisher to not use `os.gentenv("HOME")`?

Reply:
Since Windows doesn't have a `.cache` folder (cache is stored in user/AppData), I'm thinking maybe we could just create a `.boutiques` directory in the user's home dir that would store the config file with the Zenodo tokens as well as the pulled descriptors.

Reply:
or just create a `.cache` directory if it doesn't exist? It's also convenient to be able to remove everything under `.cache` without having to worry about loosing configuration.